datumaro/plugins/common_segmentation_format.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import glob
+import os.path as osp
+
+import numpy as np
+
+from datumaro.components.annotation import AnnotationType, LabelCategories, Mask, MaskCategories
+from datumaro.components.extractor import DatasetItem, Importer, SourceExtractor
+from datumaro.components.format_detection import FormatDetectionContext
+from datumaro.components.media import Image
+from datumaro.util.image import find_images
+from datumaro.util.mask_tools import generate_colormap, lazy_mask
+from datumaro.util.meta_file_util import (
+    DATASET_META_FILE,
+    has_meta_file,
+    is_meta_file,
+    parse_meta_file,
+)
+
+
+class CommonSegmentationPath:
+    MASKS_DIR = "masks"
+    IMAGES_DIR = "images"
+
+
+def make_categories(label_map=None):
+    if label_map is None:
+        return {}
+
+    categories = {}
+    label_categories = LabelCategories()
+    for label in label_map:
+        label_categories.add(label)
+    categories[AnnotationType.label] = label_categories
+
+    has_colors = any(v is not None for v in label_map.values())
+    if not has_colors:  # generate new colors
+        colormap = generate_colormap(len(label_map))
+    else:  # only copy defined colors
+        label_id = lambda label: label_categories.find(label)[0]
+        colormap = {label_id(name): (desc[0], desc[1], desc[2]) for name, desc in label_map.items()}
+    mask_categories = MaskCategories(colormap)
+    mask_categories.inverse_colormap  # pylint: disable=pointless-statement
+    categories[AnnotationType.mask] = mask_categories
+    return categories
+
+
+class CommonSegmentationExtractor(SourceExtractor):
+    def __init__(
+        self,
+        path,
+        subset=None,
+        image_prefix="",
+        mask_prefix="",
+    ):
+        if not osp.isdir(path):
+            raise FileNotFoundError("Can't read dataset directory '%s'" % path)
+
+        super().__init__(subset=subset)
+
+        self._image_prefix = image_prefix
+        self._mask_prefix = mask_prefix
+
+        if has_meta_file(path):
+            label_map = parse_meta_file(path)
+            self._categories = make_categories(label_map)
+        else:
+            meta_file = glob.glob(osp.join(path, "**", DATASET_META_FILE), recursive=True)
+            if is_meta_file(meta_file[0]):
+                label_map = parse_meta_file(meta_file[0])
+                self._categories = make_categories(label_map)
+
+        self._items = list(self._load_items(path).values())
+
+    def _load_items(self, path):
+        items = {}
+
+        image_dir = osp.join(path, CommonSegmentationPath.IMAGES_DIR)
+
+        if not osp.isdir(image_dir):
+            image_dir = glob.glob(
+                osp.join(path, "**", CommonSegmentationPath.IMAGES_DIR), recursive=True
```

Comment:
Do we really need to search for the image and annotation directories like that? I understand the nested folder support, but shouldn't the image and annotation directories at least be in the same directory as the `dataset_meta.json` file? Since we have already located the meta file at this point, we should be able to just locate the images and annotations relative to it.

Reply:
I agree. Fixed it.